In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

from statsmodels.formula.api import logit
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, plot_confusion_matrix, classification_report
from sklearn import metrics
from sklearn.feature_selection import RFE


pd.options.display.max_columns = None

### 1.	Perform pre-processing on the full dataset here: https://www.kaggle.com/sohier/calcofi 

In [3]:
# bottle dataframe
df_b = pd.read_csv('../bottle.csv')

# cast dataframe
df_c = pd.read_csv('../cast.csv')

df_b.head()

c:\python\python38\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (47,73) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
c:\python\python38\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (40,41,44,45,46,47) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


Cst_Cnt  Btl_Cnt       Sta_ID                                Depth_ID  \
0        1        1  054.0 056.0  19-4903CR-HY-060-0930-05400560-0000A-3   
1        1        2  054.0 056.0  19-4903CR-HY-060-0930-05400560-0008A-3   
2        1        3  054.0 056.0  19-4903CR-HY-060-0930-05400560-0010A-7   
3        1        4  054.0 056.0  19-4903CR-HY-060-0930-05400560-0019A-3   
4        1        5  054.0 056.0  19-4903CR-HY-060-0930-05400560-0020A-7   

   Depthm  T_degC  Salnty  O2ml_L  STheta  O2Sat  Oxy_µmol/Kg  BtlNum  RecInd  \
0       0   10.50  33.440     NaN  25.649    NaN          NaN     NaN       3   
1       8   10.46  33.440     NaN  25.656    NaN          NaN     NaN       3   
2      10   10.46  33.437     NaN  25.654    NaN          NaN     NaN       7   
3      19   10.45  33.420     NaN  25.643    NaN          NaN     NaN       3   
4      20   10.45  33.421     NaN  25.643    NaN          NaN     NaN       7   

   T_prec  T_qual  S_prec  S_qual  P_qual  O_qual  SThtaq  O2Satq  ChlorA  \
0     1.0     NaN     2.0     NaN     9.0     9.0     NaN     9.0     NaN   
1     2.0     NaN     2.0     NaN     9.0     9.0     NaN     9.0     NaN   
2     2.0     NaN     3.0     NaN     9.0     9.0     NaN     9.0     NaN   
3     2.0     NaN     2.0     NaN     9.0     9.0     NaN     9.0     NaN   
4     2.0     NaN     3.0     NaN     9.0     9.0     NaN     9.0     NaN   

   Chlqua  Phaeop  Phaqua  PO4uM  PO4q  SiO3uM  SiO3qu  NO2uM  NO2q  NO3uM  \
0     9.0     NaN     9.0    NaN   9.0     NaN     9.0    NaN   9.0    NaN   
1     9.0     NaN     9.0    NaN   9.0     NaN     9.0    NaN   9.0    NaN   
2     9.0     NaN     9.0    NaN   9.0     NaN     9.0    NaN   9.0    NaN   
3     9.0     NaN     9.0    NaN   9.0     NaN     9.0    NaN   9.0    NaN   
4     9.0     NaN     9.0    NaN   9.0     NaN     9.0    NaN   9.0    NaN   

   NO3q  NH3uM  NH3q  C14As1  C14A1p  C14A1q  C14As2  C14A2p  C14A2q  DarkAs  \
0   9.0    NaN   9.0     NaN     NaN     9.0     NaN     NaN     9.0     NaN   
1   9.0    NaN   9.0     NaN     NaN     9.0     NaN     NaN     9.0     NaN   
2   9.0    NaN   9.0     NaN     NaN     9.0     NaN     NaN     9.0     NaN   
3   9.0    NaN   9.0     NaN     NaN     9.0     NaN     NaN     9.0     NaN   
4   9.0    NaN   9.0     NaN     NaN     9.0     NaN     NaN     9.0     NaN   

   DarkAp  DarkAq  MeanAs  MeanAp  MeanAq IncTim  LightP  R_Depth  R_TEMP  \
0     NaN     9.0     NaN     NaN     9.0    NaN     NaN      0.0   10.50   
1     NaN     9.0     NaN     NaN     9.0    NaN     NaN      8.0   10.46   
2     NaN     9.0     NaN     NaN     9.0    NaN     NaN     10.0   10.46   
3     NaN     9.0     NaN     NaN     9.0    NaN     NaN     19.0   10.45   
4     NaN     9.0     NaN     NaN     9.0    NaN     NaN     20.0   10.45   

   R_POTEMP  R_SALINITY  R_SIGMA  R_SVA  R_DYNHT  R_O2  R_O2Sat  R_SIO3  \
0     10.50      33.440    25.64  233.0     0.00   NaN      NaN     NaN   
1     10.46      33.440    25.65  232.5     0.01   NaN      NaN     NaN   
2     10.46      33.437    25.65  232.8     0.02   NaN      NaN     NaN   
3     10.45      33.420    25.64  234.1     0.04   NaN      NaN     NaN   
4     10.45      33.421    25.64  234.0     0.04   NaN      NaN     NaN   

   R_PO4  R_NO3  R_NO2  R_NH4  R_CHLA  R_PHAEO  R_PRES  R_SAMP  DIC1  DIC2  \
0    NaN    NaN    NaN    NaN     NaN      NaN       0     NaN   NaN   NaN   
1    NaN    NaN    NaN    NaN     NaN      NaN       8     NaN   NaN   NaN   
2    NaN    NaN    NaN    NaN     NaN      NaN      10     NaN   NaN   NaN   
3    NaN    NaN    NaN    NaN     NaN      NaN      19     NaN   NaN   NaN   
4    NaN    NaN    NaN    NaN     NaN      NaN      20     NaN   NaN   NaN   

   TA1  TA2  pH2  pH1 DIC Quality Comment  
0  NaN  NaN  NaN  NaN                 NaN  
1  NaN  NaN  NaN  NaN                 NaN  
2  NaN  NaN  NaN  NaN                 NaN  
3  NaN  NaN  NaN  NaN                 NaN  
4  NaN  NaN  NaN  NaN                 NaN

In [4]:
df_b.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 864863 entries, 0 to 864862
Data columns (total 74 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Cst_Cnt              864863 non-null  int64  
 1   Btl_Cnt              864863 non-null  int64  
 2   Sta_ID               864863 non-null  object 
 3   Depth_ID             864863 non-null  object 
 4   Depthm               864863 non-null  int64  
 5   T_degC               853900 non-null  float64
 6   Salnty               817509 non-null  float64
 7   O2ml_L               696201 non-null  float64
 8   STheta               812174 non-null  float64
 9   O2Sat                661274 non-null  float64
 10  Oxy_µmol/Kg          661268 non-null  float64
 11  BtlNum               118667 non-null  float64
 12  RecInd               864863 non-null  int64  
 13  T_prec               853900 non-null  float64
 14  T_qual               23127 non-null   float64
 15  S_prec           

In [5]:
# Removing extra spaces from column names using str.replace() method
df_b.columns = df_b.columns.str.replace(" ","")

In [6]:
# droping columns form dataframe where not null value less thresold = 500000
df1 = df_b.dropna(axis=1, thresh=500000)

df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 864863 entries, 0 to 864862
Data columns (total 35 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Cst_Cnt      864863 non-null  int64  
 1   Btl_Cnt      864863 non-null  int64  
 2   Sta_ID       864863 non-null  object 
 3   Depth_ID     864863 non-null  object 
 4   Depthm       864863 non-null  int64  
 5   T_degC       853900 non-null  float64
 6   Salnty       817509 non-null  float64
 7   O2ml_L       696201 non-null  float64
 8   STheta       812174 non-null  float64
 9   O2Sat        661274 non-null  float64
 10  Oxy_µmol/Kg  661268 non-null  float64
 11  RecInd       864863 non-null  int64  
 12  T_prec       853900 non-null  float64
 13  S_prec       817509 non-null  float64
 14  P_qual       673755 non-null  float64
 15  Chlqua       639166 non-null  float64
 16  Phaqua       639170 non-null  float64
 17  SiO3qu       510866 non-null  float64
 18  NO2q         529474 non-

In [7]:
# droping some more columns that are redundant or not useful.
df2 = df1.drop(['Cst_Cnt','Btl_Cnt','Sta_ID','Depth_ID','T_prec','S_prec','P_qual','Chlqua','Phaqua','SiO3qu',
               'NO2q','NO3q','NH3q','C14A1q','C14A2q','DarkAq','MeanAq','Oxy_µmol/Kg','R_SALINITY','R_SIGMA',
                'O2ml_L','R_O2','R_O2Sat','R_TEMP','R_POTEMP'], axis=1)

# drop null value from salnty column
df3 = df2[df2['Salnty'].notnull()]
df4 = df3[df3['T_degC'].notnull()]

df5 = df4[df4['Salnty'].notnull()]
df6 = df5[df5['T_degC'].notnull()]
#df7 = df6[df6[:].notnull()]

df7= df6[df6['O2Sat'].notnull()]
df7 = df7[df7['STheta'].notnull()]
df8 = df7[df7['R_SVA'].notnull()]
df9 = df8[df8['R_DYNHT'].notnull()]
df9.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 655898 entries, 2160 to 864862
Data columns (total 10 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   Depthm   655898 non-null  int64  
 1   T_degC   655898 non-null  float64
 2   Salnty   655898 non-null  float64
 3   STheta   655898 non-null  float64
 4   O2Sat    655898 non-null  float64
 5   RecInd   655898 non-null  int64  
 6   R_Depth  655898 non-null  float64
 7   R_SVA    655898 non-null  float64
 8   R_DYNHT  655898 non-null  float64
 9   R_PRES   655898 non-null  int64  
dtypes: float64(7), int64(3)
memory usage: 55.0 MB


In [8]:
df9.var()

Depthm     93027.944353
T_degC        17.777238
Salnty         0.211333
STheta         1.140032
O2Sat       1373.309101
RecInd         3.876933
R_Depth    93027.944737
R_SVA       8754.651599
R_DYNHT        0.145530
R_PRES     95040.784520
dtype: float64

In [9]:
# Scaling dataset using Standard Scaler method because of too high variance for some of the columns showed above.
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

df10 = pd.DataFrame(scaler.fit_transform(df9),columns=df9.columns)

df10.var()

Depthm     1.000002
T_degC     1.000002
Salnty     1.000002
STheta     1.000002
O2Sat      1.000002
RecInd     1.000002
R_Depth    1.000002
R_SVA      1.000002
R_DYNHT    1.000002
R_PRES     1.000002
dtype: float64

In [10]:
df10.describe()

Depthm        T_degC        Salnty        STheta         O2Sat  \
count  6.558980e+05  6.558980e+05  6.558980e+05  6.558980e+05  6.558980e+05   
mean  -1.224144e-17 -4.692478e-16 -8.042341e-15  5.772844e-15 -5.854223e-17   
std    1.000001e+00  1.000001e+00  1.000001e+00  1.000001e+00  1.000001e+00   
min   -7.124651e-01 -2.252151e+00 -9.632854e+00 -4.493373e+00 -1.548304e+00   
25%   -5.518118e-01 -7.366053e-01 -7.663732e-01 -8.023379e-01 -9.681344e-01   
50%   -3.026352e-01 -1.816167e-01  4.283460e-02  1.604594e-01 -6.954679e-02   
75%    2.612907e-01  7.338773e-01  7.628555e-01  7.823440e-01  1.088093e+00   
max    1.683153e+01  4.791935e+00  6.968957e+00  1.979781e+02  4.231800e+00   

             RecInd       R_Depth         R_SVA       R_DYNHT        R_PRES  
count  6.558980e+05  6.558980e+05  6.558980e+05  6.558980e+05  6.558980e+05  
mean  -1.626053e-16 -1.367141e-17 -5.850756e-16  1.701749e-16 -4.294252e-17  
std    1.000001e+00  1.000001e+00  1.000001e+00  1.000001e+00  1.000001e+00  
min   -8.462996e-01 -7.124652e-01 -2.375648e+00 -1.137424e+00 -7.095728e-01  
25%   -8.462996e-01 -5.518118e-01 -8.312871e-01 -7.704361e-01 -5.506298e-01  
50%   -8.462996e-01 -3.026353e-01 -1.846863e-01 -2.199542e-01 -3.008622e-01  
75%    1.185197e+00  2.612907e-01  8.456000e-01  5.402350e-01  2.603039e-01  
max    1.185197e+00  1.683153e+01  4.857732e+00  9.033384e+00  1.699473e+01

### 2.	Perform feature selection (decide what columns should be included in your analysis). How did you figure out what features were important?


Based on RFE feature selection and selecting number of features 5 because we will be using 5 features to predict temperature of water, we can see from the ranking list below, columns Depthm,Salnty,R_Depth,R_SVA and R_PRES are the features we need to predict target variable T_degC. 

As you can see you in the ranking list below you get 1 as your first choices for the features for target variable temperature. And based on this result we have selected these features.


Selected Features (columns) : Depthm , Salnty , R_Depth , R_SVA , R_PRES

Target Feature (column)    : T_degC

In [11]:
from sklearn.feature_selection import RFE

X = df10.drop('T_degC',axis=1)
y = df10['T_degC']

reg = LinearRegression()
rfe = RFE(reg,n_features_to_select=5)
rfe_ = rfe.fit(X, y)

print(X.columns)
print(rfe_.support_)
print(rfe.ranking_)


Index(['Depthm', 'Salnty', 'STheta', 'O2Sat', 'RecInd', 'R_Depth', 'R_SVA',
       'R_DYNHT', 'R_PRES'],
      dtype='object')
[ True  True False False False  True  True False  True]
[1 1 5 2 4 1 1 3 1]


### 3.	Perform multiple linear regression on the dataset. Be sure to display the final resulting equation with the coefficients and intercept in markdown. 

In [12]:
reg = LinearRegression()

X = df10.drop(['STheta','T_degC','O2Sat','RecInd','R_DYNHT'],axis=1)
y = df10['T_degC']

#print(X.columns)
#print(y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,  random_state=45)

reg.fit(X_train,y_train)

accuracy_score = reg.score(X_test, y_test)

print(accuracy_score)

0.9944629622002892


In [13]:
print(reg.intercept_)
print(reg.coef_)

-1.581246576975526e-05
[  7.00202924   0.36027012 -21.245952     1.09556118  14.09500717]


#### Final Equation:

#### Target variable = Intercept + coeff_1(explanatory_var_1) + coeff_2(explanatory_var_2) + ... + coeff_x(explanatory_var_x)

Temperature (T_degC) = (-1.581246576975526e-05) + (7.00202924)*Depthm +  (0.36027012)*Salnty + (-21.245952)*R_Depth +     
                        (1.09556118)*R_SVA + (14.09500717)*R_PRES


### 4.	Which regression approach had the “best” performance? Be sure to compare it to what you did last week.


Multiple linear regression had the 'best' performance compared to linear regression with one explanatory variable. If we compare the result score of linear regression from last week to multiple linear regression this week you can see big difference in score. Last week we had score about 0.2536793464693524 with one feature and this week with 5 features we got score of 0.9944629622002892 which is a huge improvement.

One Feature Linear Regression score            :      0.2536793464693524   (25%)

Multiple features (5) Linear regression score  :      0.9944629622002892   (99%)

### 5.	Work with the diabetes dataset to perform multiple logistic regression. Look at the documentation for logistic regression in statsmodel and/or sklearn and see if you can tune the model performance based on available parameters. Please note: data scientists need to constantly learn new approaches to optimizing models. This is one way of doing that. 

In [14]:
dia_df = pd.read_csv('../diabetes.csv')

dia_df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [15]:

X = dia_df.drop('Outcome',axis=1)
y = dia_df['Outcome']

reg = LinearRegression()
rfe = RFE(reg,n_features_to_select=3)
rfe_ = rfe.fit(X, y)

print(X.columns)
print(rfe_.support_)
print(rfe.ranking_)

#Based on the ranking list from below we will be using BMI and DiabetesPedigreeFunction features as our explanatory
#variables.

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age'],
      dtype='object')
[ True False False False False  True  True False]
[1 2 4 6 5 1 1 3]


In [16]:
X = dia_df.drop('Outcome', axis=1)
y = dia_df['Outcome']

# Split into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state=42, stratify=y)

#Standardize
sc= StandardScaler()
X_train=sc.fit_transform(X_train)
X_test=sc.fit_transform(X_test)

In [17]:
log_reg = LogisticRegression()

# fit the model with data
log_reg.fit(X_train,y_train)

#
y_pred = log_reg.predict(X_test)

score = log_reg.score(X,y)
print(score)

0.3489583333333333


In [18]:
conf_matrix = confusion_matrix(y_test, y_pred)
conf_matrix

array([[105,  20],
       [ 31,  36]], dtype=int64)

In [22]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.77      0.84      0.80       125
           1       0.64      0.54      0.59        67

    accuracy                           0.73       192
   macro avg       0.71      0.69      0.69       192
weighted avg       0.73      0.73      0.73       192



In [19]:
print("Accuracy :",metrics.accuracy_score(y_test, y_pred))
print("Precision :",metrics.precision_score(y_test, y_pred))
print("Recall :",metrics.recall_score(y_test, y_pred))

Accuracy : 0.734375
Precision : 0.6428571428571429
Recall : 0.5373134328358209


In [21]:
from statsmodels.formula.api import logit

bmi_ped_int=logit("Outcome ~ BMI * DiabetesPedigreeFunction", data=dia_df).fit()
print(bmi_ped_int.params)


Optimization terminated successfully.
         Current function value: 0.590169
         Iterations 6
Intercept                      -4.026265
BMI                             0.090307
DiabetesPedigreeFunction        0.991827
BMI:DiabetesPedigreeFunction   -0.002047
dtype: float64


### 6.	What are the other ways to assess the performance of the model besides accuracy? What approach makes the most sense with the diabetes dataset? Why? 

There are also sensitivity and specificity measures we can use to assess performace of our model. Sensitivity is the proportion of true observations that are correctly predicted by the model as being true. Specificity is the proportion of false observations that are correctly predicted by the model as being false.

### 7.	Is the KNN model or the multiple logistic regression model more performant for the diabetes dataset? How can you tell? 

KNN model is better because it has higher score than multiple logistic regression

In [25]:
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

knn = KNeighborsClassifier(n_neighbors=14)

knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)

In [26]:
print(knn.score(X_test, y_test))

0.7604166666666666
